pkg/controller/servicediscovery/privatednsnamespace/hooks.go

```diff
@@ -0,0 +1,160 @@
+			cr.Status.SetConditions(xpv1.Unavailable())
+		}
+		return ret, awsclients.Wrap(cpresource.IgnoreAny(err, ActualIsNotFound, IsDuplicateRequest), "get-namespace failed")
+	}
+	ret.ResourceExists = true
+
+	cr.Status.SetConditions(xpv1.Available())
+	if externalName != *namespaceID {
+		meta.SetExternalName(cr, *namespaceID)
+		ret.ResourceLateInitialized = true
+	}
+
+	currentSpec := cr.Spec.ForProvider.DeepCopy()
+	lateInitialize(&cr.Spec.ForProvider, nsReqResp)
+	ret.ResourceLateInitialized = ret.ResourceLateInitialized || !cmp.Equal(&cr.Spec.ForProvider, currentSpec)
+
+	upToDate, err := e.isUpToDate(cr, nsReqResp)
+	if err != nil {
+		return ret, errors.Wrap(err, "isUpToDate check failed")
+	}
+
+	ret.ResourceUpToDate = upToDate
+	return ret, nil
+}
+
+func (e *external) updateImpl(ctx context.Context, mg cpresource.Managed) (managed.ExternalUpdate, error) {
+	// Update always returns an error, since the servicediscovery API does not allow updating namespaces
+	return managed.ExternalUpdate{}, nil
+}
+
+func (e *external) isUpToDate(cr *svcapitypes.PrivateDNSNamespace, resp *svcsdk.GetNamespaceOutput) (bool, error) { // nolint:unparam
+	if meta.GetExternalName(cr) != *resp.Namespace.Id {
+		return false, nil
+	}
+
+	if cr.Spec.ForProvider.CreatorRequestID != resp.Namespace.CreatorRequestId {
+		return false, nil
+	}
+	if cr.Spec.ForProvider.Description != resp.Namespace.Description {
+		return false, nil
+	}
+
+	// if cr.Spec.ForProvider.VPC
+	// VPC information is not available through servicediscovery API. Instead
+	// one could use the HostedZone information to verify the VPC configuration
+	// through Route53 HostedZone records
+
+	// tags := map[string]string
+	// if cmp.Equal(cr.Spec.ForProvider.Tags, tags) {
+	// Where does the servicediscovery API provide Tag information?
+
+	// if cr.Spec.ForProvider.Region
+	// Region information is not available through servicediscovery API
+	return true, nil
+}
+
+func preCreate(ctx context.Context, cr *svcapitypes.PrivateDNSNamespace, input *svcsdk.CreatePrivateDnsNamespaceInput) error {
+	input.Name = awsclients.String(meta.GetExternalName(cr))
+	return nil
+}
+
+func (e *external) deleteImpl(ctx context.Context, mg cpresource.Managed) error {
+	cr, ok := mg.(*svcapitypes.PrivateDNSNamespace)
+	if !ok {
+		return errors.New(errUnexpectedObject)
+	}
+	input := &svcsdk.DeleteNamespaceInput{
+		Id: awsclients.String(meta.GetExternalName(cr)),
+	}
+	_, err := e.client.DeleteNamespaceWithContext(ctx, input)
+	return awsclients.Wrap(cpresource.IgnoreAny(err, ActualIsNotFound, IsDuplicateRequest), errDelete)
+}
+
+func lateInitialize(forProvider *svcapitypes.PrivateDNSNamespaceParameters, nsReqResp *svcsdk.GetNamespaceOutput) {
+	if nsReqResp == nil {
+		return
+	}
+
+	if forProvider.Description == nil && nsReqResp.Namespace.Description != nil {
+		forProvider.Description = nsReqResp.Namespace.Description
+	}
+
+	if forProvider.CreatorRequestID == nil && nsReqResp.Namespace.CreatorRequestId != nil {
```

Comment:
I think you can remove the second nil check from both if's. `Description` and `CreatorRequestID` will be set to nil anyway.

Reply:
Again, referenced from another controller, but I agree. Updated without the second test.